torch/csrc/jit/runtime/interpreter.cpp

```diff
@@ -1495,21 +1516,36 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
             ++af.pc;
           } break;
           case WARN: {
+            // Keeps track of which WARN instruction has been executed before,
+            // we only want to execute each WARN once to match default Python
+            // warning behavior.
+            // TODO TODO this set should be graph specific, rather than global.
+            bool need_warn = true;
+            if (inst.X != -1) {
+              auto inserted = warned_indices_.insert(inst.X);
+              need_warn = inserted.second;
+            }
+
             Node* node = frames.back().function->instructions_source_.at(af.pc);
             auto range = node->sourceRange().source();
             if (range->filename()) {
-              auto line = range->starting_line_no() +
-                  range->lineno_for_offset(node->sourceRange().start());
               drop(stack, 1);
-              c10::SourceLocation location{
-                  "", range->filename()->c_str(), uint32_t(line)};
-              // Sends the warning to the warning handler with the
-              // "verbatim" flag. This flag ensures the warning handler
-              // will print the exception as configured.
-              c10::Warning::warn(
-                  location, pop(stack).toStringRef(), /*verbatim=*/true);
+              const auto msg = pop(stack).toStringRef();
+              if (need_warn) {
+                auto line = range->starting_line_no() +
+                    range->lineno_for_offset(node->sourceRange().start());
+                c10::SourceLocation location{
+                    "", range->filename()->c_str(), uint32_t(line)};
+                // Sends the warning to the warning handler with the
+                // "verbatim" flag. This flag ensures the warning handler
+                // will print the exception as configured.
+                c10::Warning::warn(location, msg, /*verbatim=*/true);
+              }
             } else {
-              TORCH_WARN(pop(stack).toStringRef());
+              const auto msg = pop(stack).toStringRef();
+              if (need_warn) {
+                TORCH_WARN(msg);
+              }
```

Comment:
it was the case with this code even earlier, but I find it suspicious that this branch didn't have `drop(stack,1)`. Do we actually have unittests that verify it? It might be a bug because I think that additional argument (i.e. stacklevel) shouldn't depend on presence or absence of range information. Safer fix might be to record # of node.inputs in `inst.N` and use it to drop from the stack

Reply:
I think adding an extra stack causes interpreter error. It kind of makes sense since lack of file info means `stack_level` is not meaningful. Anyway, I will do some more investigation and address this issue in a later PR.

Reply:
Yeah, I'm just asking whether we actually ever have a codepath exercising this branch (as usually we have the file info).